docs/design/epm/design.md

```diff
@@ -0,0 +1,206 @@
+var EnclavePoolStore map[int]*v1alpha1.Enclave
+each enclave manager will implement the following interface, currently we just have one type of enclave runtime, we implement it in enclave-cache.go
+
+4. EnclavePool interface
+
+EnclavePool represents one type of enclave pool, each kind of enclave pool need implement the EnclavePool interface.
+
+```code
+type EnclavePool interface {
+        // GetCache gets the cache by ID
+        GetCache(ID string) (*v1alpha1.Cache, error)
+        // SaveCache saves the data to a cache directory and record the cache metadata
+        SaveCache(sourcePath string, cache *v1alpha1.Cache) error
+        // SaveFinalCache save the final enclave cache info
+        SaveFinalCache(ID string) error
+        // ListCache lists part of or all of the cache metadata
+        ListCache(lastCacheID string, limit int32) ([]*v1alpha1.Cache, error)
+        // DeleteCache deletes the specified cached data and remove the corresponding cache metadata
+        DeleteCache(ID string) error
+        // LoadCache loads the specified cache data to work directory
+        LoadCache(ID, targetPath string) error
+        // GetPoolType gets the pool type of current pool
+        GetPoolType() string
+}
+```
+
+5. DefaultEnclavePool
+
+DefaultEnclavePool is the default implementation of EnclavePool
+
+```code
+type DefaultEnclavePool struct {
+        Root          string
+        Type          string
+        Enclaveinfo   map[int]*v1alpha1.Enclave
+        CacheMetadata *cache_metadata.Metadata
+}
+```
+
+### Enclave cache pool operation
+
+We just need implement two interfaces:
+
+func (d *EnclaveCacheManager) GetCache(ID string) (*v1alpha1.Cache, error)
+
+This function will marshal enclave info in cache pool, and then sendmsg enclave file descriptor to the receiving process init-runelet in container. By sendmsg, the enclave file descriptor will be dup to init-runelet process in container which will create and map the enclave by enclave info.
+
+![getcache](getcache.png)
+
+func (d *EnclaveCacheManager) SaveCache(sourcePath string, cache *v1alpha1.Cache) error
+
+SaveCache represents enclave info will be saved into EnclavePoolStore, the control flow is as below:
+
+![savecache](savecache.png)
+
+## EPM client implementation
+
+![epmclient](epmclient.png)
+
+Save enclave from EPM client to EPM service:
+1. analyze /proc/self/maps, get device /dev/sgx/enclave’s mapping info(address, size, permission, flag)
+2. query /proc/self/fd, get device /dev/sgx/enclave’s fd
+3. grpc those info above to EPM service, but enclave fd is sent by sendmsg in unix socket.
+Get enclave from EPM service to EPM client:
+1. Calling GetCache in pal_init which will be introduced in 2.4 chapter, The GetCache function will go into EPM service by gRPC. 
+2. cacheResp will be responded with cache info from EPM service to EPM client. Enclave fd will be returned by sendmsg/recvmsg. Enclave info will be composed of cacheResp and enclave fd.
+
+## Enclave Pal API
```

Comment:
It's better to quote the [Enclave Runtime PAL API Specification](https://github.com/alibaba/inclavare-containers/blob/master/rune/libenclave/internal/runtime/pal/spec.md#enclave-runtime-pal-api-specification) directly.